### ISSUE TITLE
platform.Foundation can't implement closure in Kotlin without crashing

### ISSUE THREAD

**Issue by argallo** [JetBrains/kotlin-native#1279]
```
import platform.Foundation.*
import platform.UIKit.*

open class Request {

    fun testing() {
        val url = NSURL.URLWithString(URLString = "https://www.github.com")
        val completion: (NSData?, NSURLResponse?, NSError?) -> Unit = { data, response, error ->
            print("Hello")
        }
        if(url != null) {
            val task = NSURLSession.sharedSession.dataTaskWithURL(url = url, completionHandler = completion)

            task.resume()
        }
    }

}
```

I've written some kotlin native code and have it compile to a framework for iOS. Everything is working fine until I tried to implement dataTaskWithURL which has a completion handler. I implement the handler but when I run xcode I crash although it's still able to print "hello" to the console.

![screen shot 2018-01-31 at 11 23 40 am](https://user-images.githubusercontent.com/5818297/35634305-5ec3d29c-0679-11e8-9c43-77845f53b296.png)

Anyone know why this might be happening?

**Comment by msink** [comment]
https://github.com/JetBrains/kotlin-native/blob/master/INTEROP.md#callbacks

**Comment by argallo** [comment]
@msink What should I be wrapping here? The completionHandler or dataTaskWithUrl?

**Comment by SvyatoslavScherbina** [comment]
@argallo
Kotlin/Native has some special rules restricting cross-thread object passing.
E.g. see: https://github.com/JetBrains/kotlin-native/blob/master/runtime/src/main/kotlin/konan/worker/ObjectTransfer.kt#L22

And in your case callback seems to be called from another thread. Please see this code as the example of using `NSURLSession` within the single thread:
https://github.com/JetBrains/kotlinconf-spinner/blob/master/kotlin-native/samples/fullstack/clients/ios/src/main/kotlin/platformUtils.kt#L72

**Comment by argallo** [comment]
Thanks works great!

**Comment by par128** [comment]
The link was broken, fixing. Thank you for the help, @SvyatoslavScherbina
https://github.com/Kotlin/kotlinconf-spinner/blob/master/clients/ios/src/iosMain/kotlin/platformUtils.kt

**Comment by SvyatoslavScherbina** [comment]
Please note that this code may be obsolete. Kotlin/Native is moving fast.

Consider [using ktor](https://ktor.io/clients/http-client/multiplatform.html) if you need to make HTTP requests.

**Comment by par128** [comment]
@SvyatoslavScherbina we have evaluated ktor.
It adds up 1.7 Megabytes to .apk on Android.
But we just need to get json from the server.
So I was looking for a compact solution.

**Comment by e5l** [comment]
Hi @par128, are you using kotlinx.coroutines separately or only with ktor?

**Comment by par128** [comment]
Hi, @e5l Leonid!
When I was taking the measurements we didn't use coroutines separately. So they probably were included in these 1.7 MB. Now we use them separately. Do you think I should measure once again? How much do you estimate the size of ktor adds up to an Android app?

**Comment by e5l** [comment]
We didn't measure ktor, but there was a problem with coroutines. I try to investigate, it looks like we should tune the proguard file.
